### DISCUSSION TITLE
Database health file containing critical errors and warnings

### DISCUSSION THREAD

**Discussion by lvca** [ArcadeData/arcadedb#214] [Ideas]
Some errors could be stored in some JSONL files under a new database `report` directory. This would allow auto-recovery procedures and the user could access them.

The most common error is when an index cannot be rebuilt for some reason (out of memory, out of space, etc) and ArcadeDB automatically drops it to avoid using a broken index for queries. Once dropped there is no procedure to retry recreating the index again. Also, the user couldn't see the errors/warnings in the log, so such issues are not found until the application becomes slow (no index).

I was thinking of a file like `databases/<mydb>/health.jsonl` (jsonl format is one json per line). Example:

```json
{ "level": "critical", "impact": ["performance","constraints"], 
  "description": "Unable to rebuild index XXX: Out of memory", 
  "firstOccurred": 223232, "lastOccurred": 232232, "itemsOccurred": 3 }
```

By using JSONL, both user and ArcadeDB can read and take action on that.

Also, there could be stored performance issues, like the lack of indexes with queries

```json
{ "level": "info", "impact": ["performance"], 
  "description": "Scanned type 'X' containing more than 1M of elements. Consider using an index",
   "firstOccurred": 12345, "lastOccurred": 34567, "itemsOccurred": 134 }
```

**Comment by jimmarino** [comment]
Hi @lvca,

One thing I would consider is introducing a type id. The type id can then be correlated to a taxonomy of issues. The benefit of this is it would be easier for Arcade to handle tuning. +1 on the JSON format since that can be piped to monitoring infrastructure which can parse and generate events from it.

**Reply by lvca** [reply]
Could you please provide some examples of type ids?

**Reply by jimmarino** [reply]
Sure something simple like foo:1, bar:2. Foo and bar could be a scheme such as severity level. The number is unique.

**Reply by lvca** [reply]
So do you mean a code that represents the issue, like with Oracle error codes? https://docs.oracle.com/database/121/DRDAS/error_code.htm#DRDAS513

**Reply by jimmarino** [reply]
Yeah, something like that. I use `:` but that is just a matter of preference.
